### PR TITLE
ds_prefill: Implementation of lm_head 

### PR DESCRIPTION
**Pull request by fbajraktariTT**
### Ticket
#40575

### Problem description
After the last transformer block in the sequence, the outputs are logits. Thus, we need to convert them from hidden states to token scores.

### What's changed
Implementation for LM (Language Model) head was added. Instead of the doing the matmul on the whole input, we narrow the input tensor to only the last TILE_HEIGHT of rows and then do the matmul. This is fine because we actually only care about the last logit.

There are two options to implement lm_head after narrowing the input tensor. We can first do `ttnn.all_gather` and then `ttnn.matmul`, or do `ttnn.matmul` and then `ttnn.reduce_scatter`. Here are the perf measurements for both approaches:

- 1,164.16 µs = 25 µs + 1,139 µs
- 1,357.21 µs = 1,178 µs + 179 µs

Based on the results, the first approach was implemented in this PR.

### Checklist
- <a href="https://github.com/tenstorrent/tt-metal/actions/runs/23949517153"> (Blackhole) e2e tests </a>
- <a href="https://github.com/tenstorrent/tt-metal/actions/runs/23949534571"> (Galaxy) DeepSeek Prefill tests </a>
- <a href="https://github.com/tenstorrent/tt-metal/actions/runs/23949548532"> (T3K) T3000 e2e tests </a>

### CODE REVIEW

File: models/demos/deepseek_v3_d_p/tests/pcc/test_lm_head.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        pytest.param(32, 1024, 10240, True, id="small"),
+        pytest.param(3200, DeepSeekV3Config.EMB_SIZE, DeepSeekV3Config.VOCAB_SIZE, False, id="full-no-pcc"),
+        # fmt: on
+    ],
+)
+@pytest.mark.parametrize(
+    "mesh_device, device_params, num_links, topology",
+    [
+        pytest.param(
+            (1, 4),
```
Comment (pavlepopovic) [inline models/demos/deepseek_v3_d_p/tests/pcc/test_lm_head.py:66]: why run test on 1x4 ? We're going to have at least 8x4 in case of 1 galaxy, so can we setup a test for that path?

File: models/demos/deepseek_v3_d_p/tt/tt_lm_head.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    def __init__(
+        self,
+        mesh_device,
+        emb_dim: int = DeepSeekV3Config.EMB_SIZE,
+        vocab_size: int = DeepSeekV3Config.VOCAB_SIZE,
+        torch_weights: dict = None,
+        num_links: int = 1,
+        topology: ttnn.Topology = ttnn.Topology.Ring,
+        activations_dtype=ttnn.bfloat8_b,
+        weights_dtype=ttnn.bfloat4_b,
```
Comment (pavlepopovic) [inline models/demos/deepseek_v3_d_p/tt/tt_lm_head.py]: act=bfloat16 as default pls, and weights at least bfp8_b. Considering lm-head is the last matmul, we can even do bfloat16 for weights here.

File: models/demos/deepseek_v3_d_p/tests/pcc/test_lm_head.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+import ttnn
+from models.demos.deepseek_v3_d_p.reference.deepseek_v3_config import DeepSeekV3Config
+from models.demos.deepseek_v3_d_p.tt.moe.init_helpers import extract_mesh_config
+from models.demos.deepseek_v3_d_p.tt.tt_lm_head import TtLMHead
+from tests.ttnn.utils_for_testing import assert_with_pcc
+
+# Mapping from torch dtypes to corresponding ttnn dtypes
+TORCH_TO_TTNN_DTYPE = {
+    torch.bfloat16: ttnn.bfloat16,
+    torch.float32: ttnn.float32,
```
Comment (pavlepopovic) [inline models/demos/deepseek_v3_d_p/tests/pcc/test_lm_head.py:25]: where are we using float32? can we use float16 everywhere instead?

File: models/demos/deepseek_v3_d_p/tests/pcc/test_lm_head.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+    num_devices = mesh_device.get_num_devices()
+    mesh_shape = mesh_device.shape
+    logger.debug(f"Testing with mesh_shape={mesh_shape}, num_devices={num_devices}")
+    logger.debug(f"batch_seq_len={batch_seq_len}, emb_dim={emb_dim}, vocab_size={vocab_size}")
+
+    # Create input tensor
+    mesh_config = extract_mesh_config(mesh_device)
+    dispatch_group_size = mesh_config.dispatch_group_size
+    torch_input = torch.randn(dispatch_group_size, batch_seq_len, emb_dim).to(torch_activations_dtype)
```
Comment (pavlepopovic) [inline models/demos/deepseek_v3_d_p/tests/pcc/test_lm_head.py:125]: this should be creating a [8, 3200, ..] shape ? (or in this case [1, 3200, ...] ) ?

File: models/demos/deepseek_v3_d_p/tt/tt_lm_head.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        logger.debug(f"[TtLMHead.forward] After narrow: x.shape={x.shape}")
+
+        # ========================================
+        # Step 1: All-gather x to get full emb_dim (replicated across TP axis)
+        # ========================================
+        # Input x is sharded: (dispatch_group_size/axis0, seq_len_per_chip, emb_dim/axis1)
+        # Both shared_expert and dispatch need full emb_dim, so all-gather first
+        # Only needed if there are multiple devices in TP axis (axis 1)
+        if self.mesh_device.shape[1] > 1:
+            x_full = ttnn.all_gather(
```
Comment (ipotkonjak-tt) [inline models/demos/deepseek_v3_d_p/tt/tt_lm_head.py:175]: do we know does this have better perf that matmul + reduce scatter?

Comment (fbajraktariTT) [reply models/demos/deepseek_v3_d_p/tt/tt_lm_head.py:175]: There is a perf analysis of those 2 cases in PR description.

File: models/demos/deepseek_v3_d_p/tt/tt_lm_head.py (hunk 2 above)
Comment (fbajraktariTT) [reply models/demos/deepseek_v3_d_p/tt/tt_lm_head.py]: Done. Since this is lm_head layer we can use both activations and weights in ttnn.bfloat16 dtype.

File: models/demos/deepseek_v3_d_p/tests/pcc/test_lm_head.py (hunk 3 above)
Comment (fbajraktariTT) [reply models/demos/deepseek_v3_d_p/tests/pcc/test_lm_head.py:25]: Yes, we can use bfloat16.

File: models/demos/deepseek_v3_d_p/tests/pcc/test_lm_head.py (hunk 4 above)
Comment (fbajraktariTT) [reply models/demos/deepseek_v3_d_p/tests/pcc/test_lm_head.py:125]: Yes, do you have some other suggestion?

File: models/demos/deepseek_v3_d_p/tests/pcc/test_lm_head.py (hunk 1 above)
Comment (fbajraktariTT) [reply models/demos/deepseek_v3_d_p/tests/pcc/test_lm_head.py:66]: Let's for now add (2, 2) test since it will be enough to cover the case of multi-row mesh shape and will be able to execute on QB.

File: models/demos/deepseek_v3_d_p/tests/pcc/test_lm_head.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+            marks=pytest.mark.requires_mesh_topology(mesh_shape=(1, 4), topology="ring"),
+            id="ring-4",
+        ),
+        pytest.param(
+            (2, 2),
+            {"fabric_config": ttnn.FabricConfig.FABRIC_1D_RING},
+            1,
+            ttnn.Topology.Ring,
+            marks=pytest.mark.requires_mesh_topology(mesh_shape=(2, 2), topology="ring"),
+            id="ring-4",
```
Comment (mbezuljTT) [inline models/demos/deepseek_v3_d_p/tests/pcc/test_lm_head.py]: nit: typo same id name?

File: models/demos/deepseek_v3_d_p/tests/pcc/test_lm_head.py (hunk 4 above)
Comment (pavlepopovic) [reply models/demos/deepseek_v3_d_p/tests/pcc/test_lm_head.py:125]: nope, just asking if it does that :)

**Review by pavlepopovic — APPROVED** [review]
